### PR TITLE
issue: 3382805 Fixing incorrect handling of src field of IPv4 routing 

### PR DESCRIPTION
**Pull request by AlexanderGrissik**
## Description
Fix usage of src routing field of IPv4

##### What
The src field should be used for src address selection only (As done for IPv6) and not as identifying interfaces or routing decisions.
This eliminates incorrect behavior and warnings such as "could not figure out source ip for rtv".

##### Why ?
Functionality

## Change type
What kind of change does this PR introduce?
- [X] Bugfix
- [ ] Feature
- [ ] Code style update
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Tests
- [ ] Other

## Check list
- [X] Code follows the style de facto guidelines of this project
- [X] Comments have been inserted in hard to understand places
- [ ] Documentation has been updated (if necessary)
- [ ] Test has been added (if possible)

### CODE REVIEW

File: src/core/sock/sockinfo_udp.cpp [hunk 1]
```diff
@@ -1127,31 +1127,16 @@ int sockinfo_udp::setsockopt(int __level, int __optname, __const void *__optval,
...
-            } else {
-                si_udp_logdbg("IPPROTO_IP, %s=%d.%d.%d.%d, mc_if:%d.%d.%d.%d mc_src:%d.%d.%d.%d",
-                              setsockopt_ip_opt_to_str(__optname), NIPQUAD(mc_grp), NIPQUAD(mc_if),
-                              NIPQUAD(mreqprm.imr_sourceaddr.s_addr));
+                ip_address resolved_ip;
+                resolve_if_ip(0, ip_address(mc_grp), resolved_ip);
+                mc_if = resolved_ip.get_in_addr();
             }
 
+            si_udp_logdbg("IPPROTO_IP, %s=%d.%d.%d.%d, mc_if:%d.%d.%d.%d mc_if:%d.%d.%d.%d",
```
Comment (pasis) [inline src/core/sock/sockinfo_udp.cpp:1137]: Inconsistency between format and arguments.

Also, you can use `ip_address` object to print pretty `to_str()` which is IP version independent. You already need to wrap `mc_grp` with `ip_address` for `resolve_if_ip()`.

File: src/core/sock/sockinfo_udp.cpp [hunk 2]
```diff
@@ -1127,31 +1127,16 @@ int sockinfo_udp::setsockopt(int __level, int __optname, __const void *__optval,
                 break;
             }
 
-            // Find local interface IP address
+            // Find local if for this MC ADD/DROP
```
Comment (pasis) [inline src/core/sock/sockinfo_udp.cpp:1130]: [suggestion] `local interface` instead of `local if`.

File: src/core/proto/route_table_mgr.cpp [hunk 3]
```diff
@@ -432,12 +432,14 @@ bool route_table_mgr::route_resolve(IN route_rule_table_key key, OUT route_resul
...
-            res = *p_val;
+            res.mtu = p_val->get_mtu();
+            res.if_index = p_val->get_if_index();
 
             rt_mgr_logdbg("dst ip '%s' resolved to if_index: %d, src-addr: %s, gw-addr: %s, "
                           "route-mtu: %" PRIu32,
                           dst_addr.to_str(family).c_str(), res.if_index,
-                          res.src.to_str(family).c_str(), res.gw.to_str(family).c_str(), res.mtu);
+                          p_val->get_src_addr().to_str(family).c_str(),
+                          p_val->get_gw_addr().to_str(family).c_str(), res.mtu);
```
Comment (pasis) [inline src/core/proto/route_table_mgr.cpp:324]: [suggestion] Mix of `res` and `p_val` in the arguments can lead to inconvenient changes in the future. I'd suggest to use a single object consistently (`p_val` in this case). Not critical.

**Review by pasis — COMMENTED** [review]
Review is not finished yet

File: src/core/sock/sockinfo_udp.cpp (hunk 1 above)
Comment (AlexanderGrissik) [reply src/core/sock/sockinfo_udp.cpp:1126]: Done

File: src/core/sock/sockinfo_udp.cpp (hunk 2 above)
Comment (AlexanderGrissik) [reply src/core/sock/sockinfo_udp.cpp]: Done

File: src/core/proto/route_table_mgr.cpp (hunk 3 above)
Comment (AlexanderGrissik) [reply src/core/proto/route_table_mgr.cpp]: Done

**Comment by AlexanderGrissik** [comment]
bot:retest

File: src/core/sock/sockinfo_udp.cpp [hunk 4]
```diff
@@ -1110,37 +1110,25 @@ int sockinfo_udp::setsockopt(int __level, int __optname, __const void *__optval,
...
-                mc_if = res.src.get_in_addr();
-                si_udp_logdbg(
-                    "IPPROTO_IP, %s=%d.%d.%d.%d, mc_if:INADDR_ANY (resolved to: %d.%d.%d.%d)",
-                    setsockopt_ip_opt_to_str(__optname), NIPQUAD(mc_grp), NIPQUAD(mc_if));
-            } else {
-                si_udp_logdbg("IPPROTO_IP, %s=%d.%d.%d.%d, mc_if:%d.%d.%d.%d mc_src:%d.%d.%d.%d",
-                              setsockopt_ip_opt_to_str(__optname), NIPQUAD(mc_grp), NIPQUAD(mc_if),
-                              NIPQUAD(mreqprm.imr_sourceaddr.s_addr));
+                ip_address resolved_ip;
+                resolve_if_ip(0, ip_address(mc_grp), resolved_ip);
```
Comment (iftahl) [inline src/core/sock/sockinfo_udp.cpp:1122]: see line 1565 in sockinfo_udp.cpp, we should probably use m_family there:
const auto &iparray = device_val->get_ip_array(AF_INET6);

instead AF_INET6

File: src/core/proto/route_table_mgr.cpp [hunk 5]
```diff
@@ -107,11 +104,6 @@ route_table_mgr::~route_table_mgr()
     // clear all route_entrys created in the constructor
     in_addr_route_entry_map_t::iterator iter;
 
-    while ((iter = m_rte_list_for_each_net_dev.begin()) != m_rte_list_for_each_net_dev.end()) {
```
Comment (iftahl) [inline src/core/proto/route_table_mgr.cpp:110]: probably can remove "in_addr_route_entry_map_t::iterator iter;" as well

File: src/core/sock/sockinfo_udp.cpp (hunk 4 above)
Comment (AlexanderGrissik) [reply src/core/sock/sockinfo_udp.cpp:1122]: Good catch

File: src/core/proto/route_table_mgr.cpp (hunk 5 above)
Comment (AlexanderGrissik) [reply src/core/proto/route_table_mgr.cpp:110]: Done

File: src/core/sock/sockinfo_udp.cpp (hunk 4 above)
Comment (AlexanderGrissik) [reply src/core/sock/sockinfo_udp.cpp:1122]: Done

**Review by iftahl — APPROVED** [review]

**Pull request merged**